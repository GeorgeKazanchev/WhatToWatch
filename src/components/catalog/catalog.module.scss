@import '../../mixins';

.visually-hidden {
    @include visually-hidden;
}

.catalog {
    width: 100%;
    max-width: 1300px;
    margin-bottom: 115px;
    margin-left: auto;
    margin-right: auto;
    padding: 0 75px;

    &-like-this {
        margin-bottom: 160px;
        padding-top: 30px;
    }
}

.title {
    margin: 0;
    margin-bottom: 12px;
    font-size: 22px;
    line-height: 26px;
    font-weight: 400;
    color: #dfcf77;
}

.genres {
    display: flex;
    flex-wrap: wrap;
    margin: 0;
    padding: 25px 0;
    list-style: none;
}

.genre {
    margin-right: 20px;
    margin-bottom: 20px;

    &:last-child {
        margin-right: 0;
    }
}

.genre-link {
    position: relative;
    display: block;
    padding-bottom: 15px;
    text-decoration: none;
    color: #dfcf77;
    transition: font-weight;

    &::after {
        content: '';
        position: absolute;
        z-index: 1;
        left: 0;
        right: 0;
        bottom: 0;
        display: none;
        height: 1px;
        background-color: #d9cd8d;
    }

    &:hover::after {
        display: block;
    }
}

.genre-active .genre-link {
    font-weight: bold;

    &::after {
        display: block;
    }
}

.movies {
    display: flex;
    flex-wrap: wrap;
    align-items: flex-start;
    margin-bottom: 60px;

    &:last-child {
        margin-bottom: 0;
    }
}

.more-button {
    display: block;
    width: 100%;
    padding: 29px 40px;
    font-size: 22px;
    line-height: 26px;
    text-align: center;
    color: #d9ca74;
    background: none;
    border: 1px solid rgba(217, 202, 116, 0.2);
    border-radius: 8px;
    transition: border-color 0.2s;
    cursor: pointer;

    &:hover {
        border-color: rgba(217, 202, 116, 0.5);
    }
}
