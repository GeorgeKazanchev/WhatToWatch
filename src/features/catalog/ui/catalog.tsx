import React from 'react';
import * as styles from './catalog.module.scss';
import { SmallMovieCard } from '../../../entities/small-movie-card';
import { getCatalogTitle, getGenres } from '../helpers/helpers';
import type { Film } from '../../../shared/types';

type Props = {
    films: Film[],
    isGenresShown?: boolean,
    isSimilarShown?: boolean,
    children?: React.ReactNode
};

export const Catalog: React.FC<Props> = ({ films, isGenresShown = false, isSimilarShown = false, children }) => {
    const genres = getGenres(films);

    return (
        <section className={`${styles.catalog} ${isSimilarShown ? styles.catalogLikeThis : ''}`}>
            <h2 className={`${styles.title} ${isSimilarShown ? '' : styles.visuallyHidden}`}>
                {getCatalogTitle(films, isSimilarShown)}
            </h2>

            {isGenresShown && (
                <ul className={styles.genres}>
                    <li className={`${styles.genre} ${styles.genreActive}`}>
                        <a href="#" className={styles.genreLink}>All genres</a>
                    </li>

                    {genres.map((genre) =>
                        <li key={genre} className={styles.genre}>
                            <a href="#" className={styles.genreLink}>{genre}</a>
                        </li>
                    )}
                </ul>
            )}

            <div className={styles.movies}>
                {films.map((film) =>
                    <SmallMovieCard
                        key={film.id}
                        film={film}
                    />
                )}
            </div>

            {children}
        </section>
    );
};
