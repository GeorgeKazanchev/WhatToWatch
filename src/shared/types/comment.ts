export type Comment = {
    id: number,
    author: CommentAuthor,
    rating: number,
    text: string,
    date: Date,
};

export type CommentAuthor = {
    id: number,
    name: string,
};
