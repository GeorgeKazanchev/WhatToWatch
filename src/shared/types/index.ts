export type { AuthInfo } from './auth-info';
export type { CommentPost } from './comment-post';
export type { Comment, CommentAuthor } from './comment';
export type { Film, FilmImages, FilmVideos } from './film';
export type { User } from './user';
