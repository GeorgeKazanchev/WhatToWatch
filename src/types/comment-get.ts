export type CommentGet = {
    id: number,
    user: CommentAuthor,
    rating: number,
    comment: string,
    date: Date,
};

export type CommentAuthor = {
    id: number,
    name: string,
};
