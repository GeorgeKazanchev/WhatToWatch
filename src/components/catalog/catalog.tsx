import React from 'react';
import * as styles from './catalog.module.scss';
import { SmallMovieCard } from '../small-movie-card/small-movie-card';
import { Film } from '../../types/film';
import { getGenres } from '../../helpers';

type Props = {
    films: Film[],
    isSimilarShown?: boolean
};

export const Catalog: React.FC<Props> = ({ films, isSimilarShown = false }) => {
    const genres = getGenres(films);

    return (
        <section className={`${styles.catalog} ${isSimilarShown ? styles.catalogLikeThis : ''}`}>
            <h2 className={`${styles.title} ${isSimilarShown ? '' : styles.visuallyHidden}`}>
                {isSimilarShown ? 'More like this' : 'Catalog'}
            </h2>

            {!isSimilarShown && (
                <ul className={styles.genres}>
                    <li className={`${styles.genre} ${styles.genreActive}`}>
                        <a href="#" className={styles.genreLink}>All genres</a>
                    </li>

                    {genres.map((genre) =>
                        <li key={genre} className={styles.genre}>
                            <a href="#" className={styles.genreLink}>{genre}</a>
                        </li>
                    )}
                </ul>
            )}

            <div className={styles.movies}>
                {films.map((film) =>
                    <SmallMovieCard
                        key={film.id}
                        film={film}
                    />
                )}
            </div>

            {!isSimilarShown && (
                <div>
                    <button className={styles.moreButton} type="button">Show more</button>
                </div>
            )}
        </section>
    );
};
