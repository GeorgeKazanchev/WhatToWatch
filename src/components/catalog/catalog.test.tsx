import React from 'react';
import renderer from 'react-test-renderer';
import { describe, it, expect } from '@jest/globals';
import { Catalog } from './catalog';
import { Film } from '../../types/film';

const testFilms: Film[] = [
    {
        id: 1,
        title: 'The Aviator',
        images: {
            posterSrc: '/img/poster-aviator.jpg',
            previewSrc: '/img/aviator.jpg',
            backgroundSrc: '/img/bg-aviator.jpg',
            backgroundColor: '#ffffff'
        },
        videos: {
            videoLink: '#',
            previewVideoLink: '#'
        },
        description: 'A biopic depicting the early years of legendary director and aviator Howard Hughes\' career from the late 1920s to the mid 1940s.',
        rating: 7.7,
        scoresCount: 907,
        director: 'Martin Scorsese',
        starring: [
            'Leonardo DiCaprio',
            'Cate Blanchett',
            'Kate Beckinsale'
        ],
        runtimeInMinutes: 170,
        genre: 'Biography',
        releaseYear: 2004,
        isFavorite: true
    },
    {
        id: 2,
        title: 'We Need to Talk About Kevin',
        images: {
            posterSrc: '/img/poster-we-need-to-talk-about-kevin.jpg',
            previewSrc: '/img/we-need-to-talk-about-kevin.jpg',
            backgroundSrc: '/img/bg-we-need-to-talk-about-kevin.jpg',
            backgroundColor: '#ffffff'
        },
        videos: {
            videoLink: '#',
            previewVideoLink: '#'
        },
        description: 'Kevin\'s mother struggles to love her strange child despite the increasingly dangerous things he says and does as he grows up. But Kevin is just getting started, and his final act will be beyond anything anyone imagined.',
        rating: 6.8,
        scoresCount: 522,
        director: 'Lynne Ramsay',
        starring: [
            'Tilda Swinton',
            'John C. Reilly',
            'Ezra Miller'
        ],
        runtimeInMinutes: 112,
        genre: 'Drama',
        releaseYear: 2011,
        isFavorite: false
    }
];

describe('Catalog', () => {
    it('Correctly renders', () => {
        const tree = renderer
            .create(
                <Catalog
                    films={testFilms}
                />
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
