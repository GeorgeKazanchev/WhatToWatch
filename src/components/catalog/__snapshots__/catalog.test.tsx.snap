// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog Correctly renders 1`] = `
<section
  className="undefined "
>
  <h2
    className="undefined undefined"
  >
    Catalog
  </h2>
  <ul>
    <li
      className="undefined undefined"
    >
      <a
        href="#"
      >
        All genres
      </a>
    </li>
    <li>
      <a
        href="#"
      >
        Biography
      </a>
    </li>
    <li>
      <a
        href="#"
      >
        Drama
      </a>
    </li>
  </ul>
  <div>
    <article>
      <div>
        <img
          alt="The Aviator"
          height="175"
          src="/img/aviator.jpg"
          width="280"
        />
      </div>
      <h3>
        <a
          href="/films/1"
        >
          The Aviator
        </a>
      </h3>
    </article>
    <article>
      <div>
        <img
          alt="We Need to Talk About Kevin"
          height="175"
          src="/img/we-need-to-talk-about-kevin.jpg"
          width="280"
        />
      </div>
      <h3>
        <a
          href="/films/2"
        >
          We Need to Talk About Kevin
        </a>
      </h3>
    </article>
  </div>
  <div>
    <button
      type="button"
    >
      Show more
    </button>
  </div>
</section>
`;
